backend/ibutsu_server/widgets/importance_component.py

```diff
@@ -0,0 +1,172 @@
+# from ibutsu_server.db.base import session
+# from sqlalchemy import func
+# from ibutsu_server.filters import apply_filters
+from ibutsu_server.filters import string_to_column
+# from ibutsu_server.db.base import Text
+
+#TODO: these were mocked methods, delete if unused
+"""
+def get_runs_from_job(job, additional_filters):
+    # Get a list of runs from the job, trim based on filters
+    # this can be imported/copied from the jenkins analysis view
+    pass
+
+
+def get_results_from_runs(run_list):
+    # Get a list of results from a list of runs
+    # this can be specifially tailored to return one already
+    # sorted by component, can genericize later
+    pass
+"""
+
+
+def get_importance_component(env="prod",
+                             group_field="component",
+                             job_name="",
+                             components=None,
+                             project=None):
+    # tmp parameters
+    #TODO: decide if this should be an optional parameter or hard coded
+    builds = 5
+
+    # taken from get_jenkins_line_chart in jenkins_job_analysis.py
+    run_limit = int((JJV_RUN_LIMIT / BARCHART_MAX_BUILDS) * builds)
+    jobs = get_jenkins_job_view(
+        filter_=f"job_name={job_name}",
+        page_size=builds,
+        project=project,
+        run_limit=run_limit
+    ).get("jobs")
+
+    # A list of job build numbers to filter our runs by
+    job_ids = []
+    for job in jobs:
+        job_ids.append(job["build_number"])
+
+    # query for RUN ids
+    # metadata has to have a string to column to work
+    # because it is a sqlalchemy property otherwise (AFAIK)
+    bnumdat = string_to_column("metadata.jenkins.build_number", Run)
+    run_data = Run.query.filter(
+        bnumdat.in_(job_ids),
+        Run.component.in_(components.split(','))
+    ).add_columns(
+        Run.id,
+        bnumdat.label("build_number")
+    ).all()
+
+    # get a list of the job IDs
+    run_info = {}
+    for run in run_data:
+        run_info[run.id] = run.build_number
+
+    mdat = string_to_column("metadata.importance", Result)
+    result_data = Result.query.filter(
+        Result.run_id.in_(run_info.keys()),
+        Result.component.in_(
+            components.split(','))).add_columns(
+                Result.run_id,
+                Result.component,
+                Result.id,
+                Result.result,
+                mdat.label("importance")
+    ).all()
+
+    #TODO: remove this when no longer needed for reference
+    """
+    sdatdict gonna look like
+    {"platfor_ui":
+        {"build451":
+            {"critical":
+                [results list],
+             "high":
+                [results list],
+            ...
+        },
+        {"build452":
+            {"critical":
+                ...
+            }
+        },
+     "rbac_frontend":
+        {...
+        }
+    }
+    """
+
+    """
+    This starts a (probably) over complicated bit of data maniplation
```

Comment:
I think a lot of what you're doing below could be achieved by building the dictionary you need and retrieving the data from DB like you did above but do it incrementally. I wouldn't know for sure how to do this at the moment but I think you could group results for each build by importance and then do that for each build in your components list? However if this works as is, I'm not totally against leaving it since you did a good job commenting what was going on.

Reply:
I'd definitely like to come back and simplify this as much as possible, but if it seems fine for now I might consider adding this as a TODO for later just for the sake of time.